include/alpaka/queue/QueueGenericThreadsNonBlocking.hpp

```diff
@@ -106,6 +106,15 @@ namespace alpaka
             return !((*this) == rhs);
         }
 
+        ~QueueGenericThreadsNonBlocking()
+        {
+            if(m_spQueueImpl.unique())
```

Comment:
This approach doesn't seem thread-safe, and in fact `shared_ptr::unique()` has been deprecated in c++17 and removed in c++20.

Reply:
Yes.. In misread this at first as, `use_count()` being deprecated.

And yes, it is indeed not save because the device hold `weak_ptr` to the `Impl`, which I also missed at first (otherwise it would effectively be safe because, when `use_count() == 1` in this d'tor no-one could make another copy, unless someone if holding a reference to `this` while it is being destructed ...).

This is more of a quick mockup to show how it should behave. A working way would be the canary between handle and impl mentioned [here](https://github.com/alpaka-group/alpaka/issues/1726#issuecomment-1133582736), but this requires more change, so I would like discuss what the preferred behavior is first.

One could also just has `~ConcurrentExecPool` wait for all tasks to complete instead of cancelling them, but then the non-blocking-queue's d'tor would block when the ref_count reaches zero, which is not desirable. Although, currently it would also block until the currently executing task is done....

Reply:
Naively I would expect that tasks enqueued in a non-blocking queue will eventually run, even if the queue object is destroyed (_e.g._ because it goes out of scope).
As you point out this is the behaviour we get from CUDA stream: calling `cudaStreamDestroy()` will not destroy the stream until all queued up work has completed:
> In case the device is still doing work in the stream `stream` when [cudaStreamDestroy()](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1gfda584f1788ca983cb21c5f4d2033a62) is called, the function will return immediately and the resources associated with `stream` will be released automatically once the device has completed all work in `stream`.

Would it work for the destructor of `ConcurrentExecPool()` to queue one last task that will wait for all work to complete and only _then_ completes the destruction of the threads and other resources associated to the pool ?

Reply:
That is the basic idea, it is not so simple, though:
1. It the d'tor (be it that of `ConcurrentExecPool` or the queue) waits for the tasks to complete destruction would block, which would be like `cudaStreamDestroy()` implicitly calling `cudaStreamSynchronize()` instead of returning immediately. Thus, destruction of the queue impl (mainly `ConcurrentExecPool`) must be avoided and moved into another thread.
2. There is another problem, though: if we put a task that holds the last ref to the queue into the queue and it destructs, we get the following call stack:
   1. `~QueueImpl`
   2. `~ConcurrentExecPool`
   3. `thread::join()`
 
   where that last call is the thread joining itself, which should result in:
   ```
   terminate called after throwing an instance of 'std::system_error'
      what():  Resource deadlock avoided
   ``` 
   At the moment, I do not understand why this is not happening in the tests with the current develop version. I had a close at `memBufTest` and found, that with the queue being captured, the ref-count are actually always one higher, than I expect, so the deleter never destructs the queue, though if queue is not captured, the deleter is sometimes not called, meaning the queue's refcount went to zero before the deleter's turn in the queue.

Reply:
I created a simplified version of the queue, which crashes when joining itself:
https://gist.github.com/jkelling/792de4fc15e5b44dff4cd96c21ece45a

Is there a for why this does happen in alpaka? Beside the refcount accidentally being such, that the deleter task never destructs the queue?

I also do not see a way to solve this self-join problem with a generic task. The stopper task must be something the `ConcurrentExecPool` is aware of, so it can both destruct savely (void self-join) and the terminate the thread without accessing `this` again.

Reply:
> 2\. where that last call is the thread joining itself, which should result in:

You can avoid this specific problem checking that `thread.get_id() != std::this_thread::get_id()` before calling `join()`.

Reply:
Ok, I understand why this is not crashing:
https://github.com/alpaka-group/alpaka/blob/22374dbfab3596c1bdda38fe90b4b40b9f6f43a1/include/alpaka/core/ConcurrentExecPool.hpp#L504-L524
The task is held in the main loop in that shared pointer. If the queue is empty after the deleter task the executor keeps the pointer and goes into the wait, leaking the queue. In most cases in the tests. There is an "asan" test failing, when the capture is removed without any other change, because sometimes the delete does not get executed and the buffer is leaked. The same address sanitizer does not detect the leak of the queue, maybe because there is cycle there.

If I insert a `currentTaskPackage = nullptr;` after line 512 and force the deleter task to be the last thing holding the queue by adding tome sleeps in test, then I see the expected exception.

Reply:
In the old PR https://github.com/alpaka-group/alpaka/pull/1727 which would only fix the warning, I modified the test and `ConcurrentExecPool` to catch this problem, so the can call that version "red".

Reply:
Apart from support for boost::fiber, this PR is ready.

@alpaka-group/alpaka-maintainers  please review.